### DISCUSSION TITLE
Mokapot computed group confidence score

### DISCUSSION THREAD

**Discussion by issararab** [wfondrie/mokapot#50] [General]
Hello,

I have a question about the group confidence score that mokapot computes. How does it compute the q-value?
I read through the paper that the training uses the whole set, but the grouped confidence score, which assigns the q-value, is calculated using the records within the group only.

isn't the q-value calculated as follows after sorting the new scores within the group?

q = (running decoys count) / (total running count)

**Comment by wfondrie** [comment]
Hi @issararab,

In mokapot we calculate the false discovery rate (FDR) for the PSMs accepted at a score threshold `t` as:
```
FDR = max((# decoys + 1) / (# targets), 1)
```
Where `# decoys` is the number of decoy PSMs with a score that exceeds `t`, and `# targets` is likewise the number of target PSMs with a score that exceeds `t`. The denominator considers only targets, because we do not ever consider decoys as correct in our final list of PSMs. The `+1` provides a conservative correction for when small numbers of PSMs are considered. 

The q-value is then a monotized version of these FDR estimates. Starting from the poorest score, we iterate backward through the list of scores. As we iterate, the q-value is the minimum FDR that has been observed. Below is an example to illustrate this. In the example, we consider higher scores as better:

| target/decoy | score | FDR | q-value |
|--------------|-------|-----|---------|
| target | 10 | (1/1) = 1.00  | 0.33 |
| target | 9 | (1/2) = 0.50  | 0.33 |
| target | 8 | (1/3) = 0.33  | 0.33 |
| decoy | 7 | (2/3) = 0.66 | 0.60 | 
| decoy | 6 | (3/3) = 1.00 | 0.60 |
| target | 5 | (3/4) = 0.75 | 0.60 |
| target | 4 | (3/5) = 0.60 | 0.60 |
| target | 3 | (4/6) = 0.66 | 0.66|
| decoy | 3 | (4/6) = 0.66 | 0.66 |
| decoy | 2 | (5/6) = 0.83|  0.71 |
| target | 1 |  (5/7) = 0.71 |  0.71 |

When we perform a grouped confidence estimate, the q-values are merely calculated separately for each group denoted by the grouping variable. In practice, one must strike a balance when using this feature - the most benefit is gained when one chooses groups that should have different score distributions, but also not too small. If a group contains too few PSMs, then the `+1` in the FDR calculation will have a large effect, as intended.

Did that answer your question?

**Reply by issararab** [reply]
Yes, thanks for the detailed explanation.
